sycl/doc/DeviceGlobal.md

```diff
@@ -0,0 +1,651 @@
+a function in the DPC++ runtime with the following information for each device
+global variable that is defined in the translation unit:
+
+* The (host) address of the variable.
+* The variable's string from the `sycl-unique-id` attribute.
+
+```
+namespace sycl::detail {
+namespace {
+
+__sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+  device_global_map::add(&::Foo,
+    /* same string returned from __builtin_sycl_unique_stable_id(::Foo) */);
+  device_global_map::add(&::inner::Bar,
+    /* same string returned from __builtin_sycl_unique_stable_id(::inner::Bar) */);
+}
+
+} // namespace (unnamed)
+} // namespace sycl::detail
+```
+
+Note that a SYCL application can legally call SYCL APIs even before `main()` by
+calling them from a global constructor.  However, the integration headers have
+been designed to ensure that the address of each device global variable is
+registered with the DPC++ runtime before the user's application could legally
+use the variable, even if that use occurs before `main()` executes.
+
+The user's application cannot legally use a device global variable until the
+variable's constructor has been called, otherwise the application would be
+using an unconstructed object which has undefined behavior by C++ rules.  Since
+all device globals must be defined at namespace scope, the C++ rules for the
+order of global constructors only guarantee that the device global will be
+constructed before subsequent global variables in the same translation unit.
+Therefore, a user application could reference a device global from another
+global constructor only if that global constructor is for an object defined
+*after* the device global in the same translation unit.  However, the
+integration header defines `__sycl_device_global_registerer` *before* all
+device globals in the user's translation unit.  Therefore, the address of all
+device global variables in the translation unit will be registered with the
+DPC++ runtime before any user code could legally use them.
+
+#### Handling shadowed device global variables
+
+The example above shows a simple case where the user's device global variables
+can all be uniquely referenced via fully qualified lookup (e.g.
+`::inner::Bar`).  However, it is possible for users to construct applications
+where this is not the case, for example:
+
+```
+sycl::device_global<int> FuBar;
+namespace {
+  sycl::device_global<int> FuBar;
+}
+```
+
+In this example, the `FuBar` variable in the global namespace shadows a
+variable with the same name in the unnamed namespace.  The integration footer
+can reference the variable in the global namespace as `::FuBar`, but there is
+no way to reference the variable in the unnamed namespace using fully qualified
+lookup.
+
+Such programs are still legal, though.  The integration footer can support
+cases like this by defining a temporary variable that holds the address of the
+shadowed device global:
```

Comment:
Is the suggestion to do this for every variable?  I am not sure the FE can tell which variable is shadowed and do it only for those.

Reply:
It is OK to do this for every device global variable in an unnamed namespace if you want.  (Or even for device global variables that are in other namespaces.)  There is a similar "shadowing" issue for `specialization_id`, which I think we handle already.  Do you treat all `specialization_id` variables in an unnamed namespaces as though they might be shadowed?  If so, then it makes sense to do the same thing for all `device_global` variables that are in an unnamed namespace.

Reply:
I don't see where we handle potential shadowing for spec_ids.   Could you show an example?   I know it is not quite relevant to this design, but would like to understand if this is handled already.

Reply:
Here is an example that shows shadowing for a `specialization_id` variable.  (The variable `::Foo` shadows the variable `Foo` in the unnamed namespace.)

```
#include <sycl/sycl.hpp>

constexpr sycl::specialization_id<int> Foo;
inline constexpr sycl::specialization_id<int> Bar{1};

namespace {
constexpr sycl::specialization_id<int> Foo;
}

int main() {
  sycl::queue q;

  q.single_task([=](sycl::kernel_handler kh){
    int i = kh.get_specialization_constant<Foo>();
    int j = kh.get_specialization_constant<Bar>();
  });
}
```

You can see that we currently handle this by compiling like this:

```
$ clang++ -fsycl -save-temps -v -o spec-const spec-const.cpp
```

The integration footer looks like this:

```
#include <CL/sycl/detail/defines_elementary.hpp>
__SYCL_INLINE_NAMESPACE(cl) {
namespace sycl {
namespace detail {
template<>
inline const char *get_spec_constant_symbolic_ID_impl<::Foo>() {
  return "f36a70ef3271d8ce____ZL3Foo";
}
} // namespace detail
} // namespace sycl
} // __SYCL_INLINE_NAMESPACE(cl)
__SYCL_INLINE_NAMESPACE(cl) {
namespace sycl {
namespace detail {
template<>
inline const char *get_spec_constant_symbolic_ID_impl<::Bar>() {
  return "_Z3Bar";
}
} // namespace detail
} // namespace sycl
} // __SYCL_INLINE_NAMESPACE(cl)
namespace {
namespace __sycl_detail {
static constexpr decltype(Foo) &__spec_id_shim_0() {
  return Foo;
}
} // namespace __sycl_detail 
} // namespace 
__SYCL_INLINE_NAMESPACE(cl) {
namespace sycl {
namespace detail {
template<>
inline const char *get_spec_constant_symbolic_ID_impl<::__sycl_detail::__spec_id_shim_0()>() {
  return "f36a70ef3271d8ce____ZN12_GLOBAL__N_13FooE";
}
} // namespace detail
} // namespace sycl
} // __SYCL_INLINE_NAMESPACE(cl)
#include <CL/sycl/detail/spec_const_integration.hpp>
```

The function `__spec_id_shim_0` is part of the implementation for handling shadowed variables.

I just verified that we generate the code for handling this shadowing even if the variable in the unnamed namespace is **not** actually shadowed.  Rather, it seems that we generate the `__spec_id_shim_0` function for any `specialization_id` variable that is defined in an unnamed namespace, regardless of whether it is shadowed.  You can see for yourself by changing the name of one of the `Foo` variables to something else.

Reply:
Thank you!